fedimint-api/src/db/mod.rs

```diff
@@ -59,13 +61,62 @@ pub type PrefixIter<'a> = Box<dyn Iterator<Item = Result<(Vec<u8>, Vec<u8>)>> +
 
 #[async_trait]
 pub trait IDatabase: Debug + Send + Sync {
-    async fn begin_transaction(&self, decoders: ModuleDecoderRegistry) -> DatabaseTransaction;
+    async fn begin_transaction<'a>(&'a self) -> Box<dyn IDatabaseTransaction<'a> + Send + 'a>;
 }
 
-dyn_newtype_define! {
-    /// A handle to a type-erased database implementation
-    #[derive(Clone)]
-    pub Database(Arc<IDatabase>)
+#[derive(Clone, Debug)]
+pub struct Database {
+    inner: Arc<dyn IDatabase>,
+    notifications: Arc<Notifications>,
+}
+
+impl<D: IDatabase + 'static> From<D> for Database {
+    fn from(inner: D) -> Self {
+        Database {
+            inner: Arc::new(inner),
+            notifications: Arc::new(Notifications::new()),
+        }
+    }
+}
+
+impl Database {
+    pub async fn begin_transaction(&self, decoders: ModuleDecoderRegistry) -> DatabaseTransaction {
+        DatabaseTransaction::new(
+            self.inner.begin_transaction().await,
+            decoders,
+            self.notifications.clone(),
+        )
+    }
+
+    // TODO: is Hash a good idea? should we call to_bytes, then hash
+    pub async fn wait_key<K>(&self, key: K, decoders: ModuleDecoderRegistry) -> Result<K::Value>
+    where
+        K: DatabaseKey + DatabaseKeyPrefixConst + Hash,
```

Comment:
is `Hash` a good idea?
I am even thinking `DatabaseKey: Hash`

Reply:
Could we just hash the serialized data? I assume when the events are saved in the transaction we have them in serialized form for the DB operation anyway. When creating the `wait_key` future here serializing is probably not too expensive.